core/dispatcher/src/main/scala/whisk/core/dispatcher/Dispatcher.scala

```diff
@@ -165,41 +134,56 @@ trait MessageDispatcher extends Registrar with Logging {
         s"failed processing message: $msg $e${e.getStackTrace.mkString("", " ", "")}"
 
     private def errorMsg(e: Throwable): String = {
-        if (e.isInstanceOf[java.util.concurrent.ExecutionException])
+        if (e.isInstanceOf[java.util.concurrent.ExecutionException]) {
             s"$e${e.getCause.getStackTrace.mkString("", " ", "")}"
-        else
+        } else {
             s"$e${e.getStackTrace.mkString("", " ", "")}"
+        }
     }
 
     private val counter = new Counter()
-    private var started = false
 }
 
-/**
- * Creates a dispatcher that uses kafka as the message pub/sub connector.
- * This is currently used by invoker.
- *
- * @param config with zookeeper host properties
- * @topic the topic to subscribe to (creates a consumer client for this topic)
- * @groupid the groupid for the topic consumer
- */
-class Dispatcher(
-    config: WhiskConfig,
-    topic: String,
-    groupid: String)(implicit val executionContext: ExecutionContext)
-    extends KafkaConsumerConnector(config.kafkaHost, groupid, topic)
-    with MessageDispatcher {
-}
+trait Registrar {
+    /**
+     * Adds handler for a message. The handler name must be unique, else
+     * the new handler replaces a previously added one unless this behavior
+     * is overridden by setting replace to false.
+     *
+     * @param handler is the message handler to add override
+     * @param replace indicates whether a new handler should replace an older handler by the same name
+     * @return an option dispatch rule, the previous value of the rule if any
+     */
+    def addHandler(handler: DispatchRule, replace: Boolean): Option[DispatchRule] = {
+        if (handler != null && handler.isValid) {
+            if (replace) handlers.put(handler.name, handler)
+            else handlers.putIfAbsent(handler.name, handler)
+        } else None
+    }
```

Comment:
note: this is NOT new code; i just moved it down in the file.

that said, historically this was used in the activator which added and remove listeners (for rule matching) dynamically.

we have a dispatcher with exactly one handler - which is why i made the comment above that we should flatten "this" (class) further.